### PR TITLE
Add cudaEvent support to the profiler

### PR DESCRIPTION
**Pull request by zdevito**
This adds the ability to record cuda timings using cudaEventRecord
in the profiler. Since it doesn't require nvprof it is easier
to run than the nvprof path.

This also records a thread id for each event, which will make
tracing results easier to understand

### CODE REVIEW

File: torch/csrc/autograd/profiler.cpp [hunk 1]
```diff
@@ -4,26 +4,30 @@
...
 #ifndef WITH_CUDA
-  if (use_cuda)
+  if (use_nvprof)
     throw std::runtime_error("Can't use CUDA profiler - PyTorch was compiled without CUDA");
 #endif
   if (profiling) {
-    if (use_cuda != using_cuda)
-      throw std::runtime_error("can't change use_cuda flag while profiler is running");
+    if (use_nvprof != using_nvprof)
+      throw std::runtime_error("can't change use_nvprof flag while profiler is running");
```
Comment (apaszke) [inline torch/csrc/autograd/profiler.cpp]: Same for `use_cuda`, right?

File: torch/csrc/autograd/profiler.h [hunk 2]
```diff
@@ -24,16 +29,93 @@ constexpr inline std::size_t ceilToMultiple(std::size_t a, std::size_t b) {
...
+#ifdef WITH_CUDA
+static void cudaCheck(cudaError_t result, const char * file, int line) {
+  if(result != cudaSuccess) {
+    std::stringstream ss;
+    ss << file << ":" << line << ": " << cudaGetErrorString(result);
+    throw std::runtime_error(ss.str());
+  }
+}
+#define PROFILER_CUDA_CHECK(result) cudaCheck(result,__FILE__,__LINE__);
+#endif
```
Comment (apaszke) [inline torch/csrc/autograd/profiler.h]: Can we have a single macro for this? There are equivalent definitions sprinkled across the codebase. Can you move it to `torch/csrc/Exceptions.h`?

File: torch/csrc/autograd/profiler.h [hunk 3]
```diff
@@ -24,16 +29,93 @@ constexpr inline std::size_t ceilToMultiple(std::size_t a, std::size_t b) {
...
 
-// NOTE: we don't need a flag saying if an event is a kernel, because it's
-// used only for the CPU-side perf recording.
-using Event = std::tuple<std::string, uint64_t, EventKind>; // (name, time, kind)
+struct Event {
+  Event(EventKind kind, std::string name, uint32_t thread_id, bool record_cuda)
+  : kind_(kind),
+  name_(std::move(name)),
+  thread_id_(thread_id),
+  cpu_ns(getTime()) {
```
Comment (apaszke) [inline torch/csrc/autograd/profiler.h]: can you please put the colon at the end of the line like you do with commas, or put all the commas at the beginning?

Comment (apaszke) [reply torch/csrc/autograd/profiler.h]: Also, can you please use trailing underscores either for all or no members?

File: torch/csrc/autograd/profiler.h [hunk 4]
```diff
@@ -24,16 +29,93 @@ constexpr inline std::size_t ceilToMultiple(std::size_t a, std::size_t b) {
...
+  }
+}
+#define PROFILER_CUDA_CHECK(result) cudaCheck(result,__FILE__,__LINE__);
+#endif
 
-// NOTE: we don't need a flag saying if an event is a kernel, because it's
-// used only for the CPU-side perf recording.
-using Event = std::tuple<std::string, uint64_t, EventKind>; // (name, time, kind)
+struct Event {
+  Event(EventKind kind, std::string name, uint32_t thread_id, bool record_cuda)
```
Comment (apaszke) [inline torch/csrc/autograd/profiler.h]: You're always passing the global value of `record_cuda` and thread-local of `thread_id`. You could just read them from inside the constructor.

File: torch/csrc/autograd/profiler.h [hunk 5]
```diff
@@ -71,59 +153,58 @@ struct RangeEventList {
 };
 
 extern bool profiling;
+extern bool using_nvprof;
```
Comment (apaszke) [inline torch/csrc/autograd/profiler.h]: Can you add a comment saying that `using_nvprof == true` doesn't imply `using_cuda == true` and that they are two distinct modes? Maybe we should just use an enum for profiler state?
```cpp
class enum ProfilerState {
  disabled,
  CPU,          // CPU-only profiling
  CUDA,         // CPU + CUDA events
  NVTX          // only emit NVTX markers
};
```

File: torch/autograd/profiler.py [hunk 6]
```diff
@@ -414,8 +465,9 @@ def parse_nvprof_trace(path):
         unique.see(row['marker_id'])
         evt = FunctionEvent(id=row['marker_id'],
                             name=strings[row['name']],
-                            start=row['start_time'],
-                            end=row['end_time'])
+                            cpu_interval=Interval(row['start_time'], row['end_time']),
+                            cuda_interval=None,
+                            thread=0)
```
Comment (apaszke) [inline torch/autograd/profiler.py]: You can get a thread id from these tables. Just look for an appropriate column (it's an sqlite db so you can open it in terminal)

File: torch/autograd/profiler.py [hunk 7]
```diff
@@ -339,37 +391,36 @@ def __missing__(self, key):
...
-    for event in functions:
-        event.start -= start_time
-        event.end -= start_time
-
-    functions.sort(key=lambda evt: evt.start)
+        elif record.kind() == 'pop':
+            function_id, start = record_stack.pop()
+            cpu_interval = Interval(
+                start_record.cpu_elapsed_us(start),
+                start_record.cpu_elapsed_us(record))
```
Comment (apaszke) [inline torch/autograd/profiler.py]: This calling convention is reversed compared to that of `torch.cuda.Event` (where `elapsed_time` is called on the end event). Can you please reverse it?

File: torch/autograd/profiler.py [hunk 8]
```diff
@@ -339,37 +391,36 @@ def __missing__(self, key):
 ################################################################################
 # CPU checkpoints
 
-Record = namedtuple('Record', ['name', 'timestamp', 'kind'])
-
-
 def parse_cpu_trace(thread_records):
     next_id = 0
-    start_time = None
+    start_record = None
```
Comment (apaszke) [inline torch/autograd/profiler.py]: This code assumes that `start_record` will be found as the first one, which might not always be the case (it's not guaranteed to be in the first list). This is why the code previously used two loops, to subtract the start time only after all events were processed.

**Review by apaszke — COMMENTED** [review]
Mostly looks good, but I have a few comments.

File: torch/autograd/profiler.py [hunk 9]
```diff
@@ -97,6 +133,10 @@ class profile(object):
         enabled (bool, optional): Setting this to False makes this context manager a no-op.
             Default: ``True``.
 
+        use_cuda (bool, optional): Enables timing of CUDA events as well using the cudaEvent API.
+            Adds approximately 4us of overhead to each tensor operation.
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/autograd/profiler.py]: It is probably about the same for recording timing, but it gives us more control over what we time. E.g. we won't time all individual kernels or cuda API calls, so I expect the overhead will be a bit lower. Really though this is to make it possible to get traces with GPU numbers without needing to export/import them.

File: torch/autograd/profiler.py (hunk 8 above)
Comment (zdevito) [reply torch/autograd/profiler.py]: Why is it not guaranteed to be first? That deserves a comment at least.

File: torch/autograd/profiler.py (hunk 7 above)
Comment (zdevito) [reply torch/autograd/profiler.py]: I think it is the same:

```def elapsed_time(self, end_event):```

Which is the same as cudaEventElapsedTime.

File: torch/autograd/profiler.py [hunk 10]
```diff
@@ -339,37 +391,36 @@ def __missing__(self, key):
...
-        event.end -= start_time
-
-    functions.sort(key=lambda evt: evt.start)
+        elif record.kind() == 'pop':
+            function_id, start = record_stack.pop()
+            cpu_interval = Interval(
+                start_record.cpu_elapsed_us(start),
+                start_record.cpu_elapsed_us(record))
+            cuda_interval = None
+            if start_record.has_cuda() and record.has_cuda():
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/autograd/profiler.py]: I am not sure, actually.The threads access use_cuda and use_nvprof without locks, so I don't have a great mental model of what is intended here. I can add an assert in any case though.

File: torch/csrc/autograd/profiler.h (hunk 4 above)
Comment (zdevito) [reply torch/csrc/autograd/profiler.h]: Having them as ctor arguments nicely decouples Event from the profiler. Otherwise if we move thread_id/use_cuda around we'd have to change this stuff too.

File: torch/autograd/profiler.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/autograd/profiler.py]: The mental model I had was that you have a single thread that enables and disables the profiler, and you don't change the flags in the middle. This would mean that either all or no events have cuda times.

File: torch/autograd/profiler.py (hunk 8 above)
Comment (apaszke) [reply torch/autograd/profiler.py]: It is first in one of these lists, but it doesn't have to be in the first one. So far it happens to be like it, because it's the main Python thread and it gets registered as the first. However, if you were to run one profile in the main thread, then spawn another one and start profiling there, the start event wouldn't appear in the first list.

File: torch/autograd/profiler.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/autograd/profiler.py]: What happens if a thread launched before or during the profiling enable outlives the end of profiling? It is going to keep push events and will race with the use_cuda flags.

File: torch/autograd/profiler.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/autograd/profiler.py]: I didn't mean to say that we shouldn't merge this because of overhead. It's just that if you consider that most of our ATen calls take 1-4us, then extra 4us really is a lot.

File: torch/autograd/profiler.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/autograd/profiler.py]: Why is it going to keep pushing events? Flag that says whether the profiling is enabled is global so it should stop queueing events once it's off, even if it outlives the profiled scope (e.g. engine threads already do this).

File: torch/autograd/profiler.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/autograd/profiler.py]: Yeah, this is for debugging convnet performance, so those overheads are more acceptable.

[the remark this answers is not in the document]
Comment (apaszke) [reply torch/autograd/profiler.py]: Sure. That still makes sense in some cases

[the remark this answers is not in the document]
Comment (apaszke) [reply torch/autograd/profiler.py]: BTW 4us is a bit abstract for most people, except for core devs who know how much 4us is. Maybe at least add a bit of context (mention avg. cost of autograd operation)?

File: setup.py [hunk 11]
```diff
@@ -542,7 +542,7 @@ def check_file(f):
     cuda_include_path = os.path.join(CUDA_HOME, 'include')
     include_dirs.append(cuda_include_path)
     include_dirs.append(tmp_install_path + "/include/THCUNN")
-    extra_compile_args += ['-DWITH_CUDA']
+    extra_compile_args += ['-DWITH_CUDA', '-DAT_CUDA_ENABLED']
```
Comment (apaszke) [inline setup.py]: btw how is this related to this PR?

File: torch/csrc/autograd/profiler.cpp [hunk 12]
```diff
@@ -16,29 +14,27 @@ void RecordFunction::pushFunctionRange(Function* fn) {
...
-  if (use_nvprof)
-    throw std::runtime_error("Can't use CUDA profiler - PyTorch was compiled without CUDA");
+  if (new_state == ProfilerState::NVTX)
+    throw std::runtime_error("Can't use NVTX profiler - PyTorch was compiled without CUDA");
 #endif
-  if (profiling) {
-    if (use_nvprof != using_nvprof)
-      throw std::runtime_error("can't change use_nvprof flag while profiler is running");
-    return;
+  if (state != ProfilerState::Disabled && new_state != state) {
```
Comment (apaszke) [inline torch/csrc/autograd/profiler.cpp]: I don't think we want to allow using `enableProfiler` to disable it. Can you please make it an error?

**Review by apaszke — COMMENTED** [review]
Two minor things and should be good to go. If you want to merge this PR without squashing please squash all but the last commit into the first one.

File: setup.py (hunk 11 above)
Comment (zdevito) [reply setup.py]: Somehow AT_CUDA_ENABLED got added into ATen headers so it has to be defined if you are including them. I have to go stamp that out, the headers shouldn't depend on cuda.

Comment (apaszke) [reply setup.py]: What do you mean by "it got added to ATen headers"? We're adding it unconditionally and this is quite worrying

Comment (zdevito) [reply setup.py]: I mean that someone made the headers depend on AT_CUDA_ENABLED being set or not. This is not good, because consumers of AT need to set it in the same way that ATen was built with. We need to remove AT_CUDA_ENABLED from all ATen public headers.

For now, I set it with WITH_CUDA so that it is at least correct for now.

Comment (ezyang) [reply setup.py]: See also https://github.com/zdevito/ATen/issues/168; I have a fix for this in my giant CuDNN refactor PR (and it might even be a single, extricable commit!)
